browser/brave_news/brave_news_tab_helper_browsertest.cc

```diff
@@ -0,0 +1,173 @@
+  }
+
+  void TearDownInProcessBrowserTestFixture() override {
+    InProcessBrowserTest::TearDownInProcessBrowserTestFixture();
+    cert_verifier_.TearDownInProcessBrowserTestFixture();
+  }
+
+  void SetUpCommandLine(base::CommandLine* command_line) override {
+    command_line->AppendSwitch(switches::kAllowRunningInsecureContent);
+    cert_verifier_.SetUpCommandLine(command_line);
+  }
+
+  void SetUpOnMainThread() override {
+    InProcessBrowserTest::SetUpOnMainThread();
+    host_resolver()->AddRule("*", "127.0.0.1");
+    cert_verifier_.mock_cert_verifier()->set_default_result(net::OK);
+  }
+
+  content::WebContents* contents() {
+    return browser()->tab_strip_model()->GetActiveWebContents();
+  }
+
+  net::EmbeddedTestServer* https_server() { return &https_server_; }
+
+ private:
+  base::test::ScopedFeatureList features_;
+  net::EmbeddedTestServer https_server_;
+  content::ContentMockCertVerifier cert_verifier_;
+};
+
+IN_PROC_BROWSER_TEST_F(BraveNewsTabHelperTest, TabHelperIsCreated) {
+  EXPECT_NE(nullptr, BraveNewsTabHelper::FromWebContents(contents()));
+}
+
+IN_PROC_BROWSER_TEST_F(BraveNewsTabHelperTest,
+                       TabHelperNotifiesObserversWhenFoundFeeds) {
+  auto* tab_helper = BraveNewsTabHelper::FromWebContents(contents());
+  WaitForFeedsChanged waiter(tab_helper);
+
+  std::vector<brave_news::mojom::FeedSearchResultItemPtr> feeds;
+  auto feed1 = brave_news::mojom::FeedSearchResultItem::New();
+  feed1->feed_url = GURL("https://example.com/1");
+
+  auto feed2 = brave_news::mojom::FeedSearchResultItem::New();
+  feed2->feed_url = GURL("https://example.com/2");
+  feeds.emplace_back(std::move(feed1));
+  feeds.emplace_back(std::move(feed2));
+  tab_helper->OnFoundFeeds(contents()->GetLastCommittedURL(), std::move(feeds));
+
+  auto result = waiter.WaitForChange();
+  EXPECT_EQ(2u, result.size());
+}
+
+IN_PROC_BROWSER_TEST_F(BraveNewsTabHelperTest, FeedsAreDeduplicated) {
+  auto* tab_helper = BraveNewsTabHelper::FromWebContents(contents());
+  WaitForFeedsChanged waiter(tab_helper);
+
+  std::vector<brave_news::mojom::FeedSearchResultItemPtr> feeds;
+  auto feed1 = brave_news::mojom::FeedSearchResultItem::New();
+  feed1->feed_url = GURL("https://example.com/1");
+  auto feed1_copy = feed1.Clone();
+  feeds.emplace_back(std::move(feed1));
+  feeds.emplace_back(std::move(feed1_copy));
+  tab_helper->OnFoundFeeds(contents()->GetLastCommittedURL(), std::move(feeds));
+
+  auto result = waiter.WaitForChange();
+  EXPECT_EQ(1u, result.size());
+}
+
+IN_PROC_BROWSER_TEST_F(BraveNewsTabHelperTest, FeedsAreFoundWhenTheyExists) {
+  ASSERT_TRUE(https_server()->Start());
+  GURL rss_page_url = https_server()->GetURL("/page_with_rss.html");
+
+  auto* tab_helper = BraveNewsTabHelper::FromWebContents(contents());
+  content::TestNavigationObserver observer(
+      contents(), content::MessageLoopRunner::QuitMode::DEFERRED);
+  NavigateParams params(browser(), rss_page_url, ui::PAGE_TRANSITION_LINK);
```

Comment:
is the `PAGE_TRANSITION_LINK` required along with `TestNavigationObserver`? wouldn't `NavigateToURLWithDisposition` with `BROWSER_TEST_WAIT_FOR_LOAD_STOP` be enough?

Reply:
also this test has failed on windows.
https://ci.brave.com/job/brave-core-build-pr-windows/job/PR-14077/25/

Reply:
Cool, I didn't know about that utility :+1: